p2p/ipld/read.go

```diff
@@ -30,23 +33,32 @@ func RetrieveBlockData(
 ) (types.Data, error) {
 	edsWidth := len(dah.RowsRoots)
 	sc := newshareCounter(ctx, uint32(edsWidth))
+	sem := semaphore.NewWeighted(maxGoRoutines)
 
 	// convert the row and col roots into Cids
 	rowRoots := dah.RowsRoots.Bytes()
 	colRoots := dah.ColumnRoots.Bytes()
 
 	// sample 1/4 of the total extended square by sampling half of the leaves in
 	// half of the rows
-	for _, row := range uniqueRandNumbers(edsWidth/2, edsWidth) {
-		for _, col := range uniqueRandNumbers(edsWidth/2, edsWidth) {
-			rootCid, err := plugin.CidFromNamespacedSha256(rowRoots[row])
-			if err != nil {
-				return types.Data{}, err
+	go func() {
+		for _, row := range uniqueRandNumbers(edsWidth/2, edsWidth) {
+			for _, col := range uniqueRandNumbers(edsWidth/2, edsWidth) {
+				rootCid, err := plugin.CidFromNamespacedSha256(rowRoots[row])
+				if err != nil {
+					select {
+					case <-sc.ctx.Done():
+					case sc.errc <- err:
+					}
+				}
+				sem.Acquire(ctx, 1)
+				go func(rootCid cid.Cid, row uint32, col uint32) {
+					sc.retrieveShare(rootCid, true, row, col, dag)
+					sem.Release(1)
+				}(rootCid, row, col)
 			}
-
-			go sc.retrieveShare(rootCid, true, row, col, dag)
 		}
-	}
+	}()
```

Comment:
I feel there are ways to achieve the same more idiomatically (without using x/sync/semaphore).